code/assurance/src/main/java/com/adobe/marketing/mobile/assurance/EventStitcher.kt

```diff
@@ -0,0 +1,131 @@
+import org.json.JSONObject
+
+internal class EventStitcher {
+
+    companion object {
+        const val LOG_TAG = "EventStitcher"
+    }
+
+    /**
+     * Queue of events that are yet to be stitched i.e waiting for the final chunk. Stored as a
+     * map with chunkId as the key and list of events as the value.
+     */
+    private val queue: MutableMap<String, MutableList<AssuranceEvent>>
+
+    /**
+     * A callback to notify the caller with the processed event (stitched or not stitched).
+     */
+    private val notifier: AdobeCallback<AssuranceEvent>
+
+    constructor(notifier: AdobeCallback<AssuranceEvent>) : this(mutableMapOf(), notifier)
+
+    constructor(queue: MutableMap<String, MutableList<AssuranceEvent>>, notifier: AdobeCallback<AssuranceEvent>) {
+        this.queue = queue
+        this.notifier = notifier
+    }
+
+    /**
+     * Receives AssuranceEvents and stitches them if they are chunked and notifies the notifier with the stitched event.
+     * If the event is not chunked, it is directly notified via the notifier.
+     *
+     * @param event the AssuranceEvent
+     */
+    @JvmName("onEvent")
+    internal fun onEvent(event: AssuranceEvent) {
+        if (!isChunked(event)) {
+            notifier.call(event)
+            return
+        }
+
+        val chunkId: String = event.metadata[AssuranceConstants.AssuranceEventKeys.CHUNK_ID] as String? ?: return
+        val totalChunks: Int = event.metadata[AssuranceConstants.AssuranceEventKeys.CHUNK_TOTAL] as Int? ?: return
+
+        // check if there are prior chunks associated with this chunkId in the queue, if not create a new queue
+        val chunkedEventsForId: MutableList<AssuranceEvent> = queue[chunkId] ?: mutableListOf()
+        chunkedEventsForId.add(event)
+
+        // if all chunks are received, stitch the events and notify the caller
+        if (chunkedEventsForId.size == totalChunks) {
+            val result: Response<AssuranceEvent, Exception> = stitch(chunkedEventsForId)
+            when (result) {
+                is Response.Success -> notifier.call(result.data)
+                is Response.Failure -> Log.error(Assurance.LOG_TAG, LOG_TAG, "Failed to stitch events for chunkId: $chunkId due to: ${result.error.message}")
+            }
+            // remove the event from the queue after stitching
+            queue.remove(chunkId)
+        } else {
+            // else add the event to the queue for stitching once all pieces are received
+            queue[chunkId] = chunkedEventsForId
+        }
+    }
+
+    /**
+     * Checks whether the event is chunked.
+     * @param event the event to check
+     * @return true if the event is chunked, false otherwise
+     */
+    @JvmName("isChunked")
+    internal fun isChunked(event: AssuranceEvent): Boolean {
+        event.metadata?.let {
+            val chunkId = it[AssuranceConstants.AssuranceEventKeys.CHUNK_ID] as String?
+            return chunkId != null
+        } ?: return false
```

Comment:
nit: Generally, I'm not inclined towards this pattern because it has the potential to execute both blocks.